lib/webview-helpers.js

```diff
@@ -0,0 +1,111 @@
+import _ from 'lodash';
+import logger from './logger';
+import { asyncMap } from 'asyncbox';
+
+const NATIVE_WIN = "NATIVE_APP";
+const WEBVIEW_WIN = "WEBVIEW";
+const WEBVIEW_BASE = WEBVIEW_WIN + "_";
+const CHROMIUM_WIN = "CHROMIUM";
+
+let helpers = {};
+
+// This function gets a list of android system processes and returns ones
+// that look like webviews, with the appropriate webview prefix and their PID
+async function webviewsFromProcs (adb, deviceSocket) {
+  let webviews = [];
+  let out = await adb.shell("cat", ["/proc/net/unix"]);
+  for (let line of out.split("\n")) {
+    line = line.trim();
+    let webviewPid = line.match(/@?webview_devtools_remote_(\d+)/);
+    if (deviceSocket) {
+      if (line.indexOf("@" + deviceSocket) === line.length - deviceSocket.length - 1) {
+        if (webviewPid) {
+          webviews.push(WEBVIEW_BASE + webviewPid[1]);
+        } else {
+          webviews.push(CHROMIUM_WIN);
+        }
+      }
+    } else if (webviewPid) {
+      // for multiple webviews a list of 'WEBVIEW_<index>' will be returned
+      // where <index> is zero based (same is in selendroid)
+      webviews.push(WEBVIEW_BASE + webviewPid[1]);
+    }
+  }
+  return _.uniq(webviews);
+}
+
+// Take a webview name like WEBVIEW_4296 and use 'adb shell ps' to figure out
+// which app package is associated with that webview. One of the reasons we
+// want to do this is to make sure we're listing webviews for the actual AUT,
+// not some other running app
+async function procFromWebview (adb, webview) {
+  // webview_devtools_remote_4296 => 4296
+  let pid = webview.match(/\d+$/);
+  if (!pid) {
+    throw new Error(`Could not find PID for webview ${webview}`);
+  }
+  pid = pid[0];
+  logger.debug(`${webview} mapped to pid ${pid}`);
+  logger.debug("Getting process name for webview");
+  let out = await adb.shell("ps");
+  let pkg = "unknown";
+  let lines = out.split(/\r?\n/);
+  /* Output of ps is like:
+   USER     PID   PPID  VSIZE  RSS     WCHAN    PC         NAME
+   u0_a136   6248  179   946000 48144 ffffffff 4005903e R com.example.test
+  */
+  let header = lines[0].trim().split(/\s+/);
+  // the column order may not be identical on all androids
+  // dynamically locate the pid and name column.
+  let pidColumn = header.indexOf("PID");
+  let pkgColumn = header.indexOf("NAME") + 1;
+
+  for (let line of lines) {
+    line = line.trim().split(/\s+/);
+    if (line[pidColumn].indexOf(pid) !== -1) {
+      logger.debug(`Parsed pid: ${line[pidColumn]} pkg: ${line[pkgColumn]}`);
+      logger.debug(`from: ${line}`);
+      pkg = line[pkgColumn];
+      break;
+    }
+  }
+  logger.debug(`returning process name: ${pkg}`);
+  return pkg;
+}
+
+helpers.getWebviews = async function (adb, deviceSocket) {
+  logger.debug("Getting a list of available webviews");
+  let webviews = await webviewsFromProcs(adb, deviceSocket);
+
+  if (deviceSocket) {
+    return webviews;
```

Comment:
?
Can we have a comment for this function, what's the 'deviceSocket' parameter for, and why do we just return here if we supplied it?